synapse/handlers/message.py

```diff
@@ -1001,13 +1001,45 @@ async def create_new_client_event(
             )
 
         self.validator.validate_new(event, self.config)
+        await self._validate_event_relation(event)
+        logger.debug("Created event %s", event.event_id)
+
+        return event, context
+
+    async def _validate_event_relation(self, event: EventBase) -> None:
+        """
+        Ensure the relation data on a new event is not bogus.
+
+        Args:
+            event: The event being created.
+
+        Raises:
+            SynapseError if the event is invalid.
+        """
+
+        relation = event.content.get("m.relates_to")
+        if not relation:
+            return
+
+        relation_type = relation.get("rel_type")
+        if not relation_type:
+            return
+
+        # Ensure the parent is real.
+        relates_to = relation["event_id"]
+        parent_event = await self.store.get_event(relates_to)
+        if not parent_event:
+            # TODO Attempt to fetch the event over federation.
+            raise SynapseError(400, "Can't send relation to unknown event")
```

Comment:
What do we do in this case? It is (theoretically) possible the user's homeserver doesn't have this backfilled *or* that the user doesn't have permission to see this event. Maybe if we can't find the event we should just skip some of the next validation and assume that clients will handle it OK?

Reply:
There was a bit of a discussion around this in #synapse-dev:matrix.org (see https://matrix.to/#/!XaqDhxuTIlvldquJaV:matrix.org/$N7UHzQ6lrerKq5LtqaPyjZmAWyP6GPCn3bMwU2BcERA?via=matrix.org&via=vector.modular.im&via=envs.net), the overall conclusion was that servers and clients must be robust to not being able to have an event for some reason so it is *probably* reasonable to skip validation if we don't know about the parent event for some reason. We might also want to see if there are other events using that as a parent relation and deduce that it is *likely* OK.